### DISCUSSION TITLE
How to incorporate Flamingo Baryon Response Emulator

### DISCUSSION THREAD

**Discussion by gcanasherrera** [cloe-org/cloelib#45] [Ideas]
Discussion board with ideas, approaches and requests to include Flamingo Baryon Response Emulator in cloelite.

**Comment by gcanasherrera** [comment]
CC @MatthieuSchaller

**Comment by MatthieuSchaller** [comment]
A couple of questions raised by our chat earlier:

- What do we want at z > larger z the emulator was trained at?

- What do we want at k > larger than where the emulator is trained?  (error, constant, or 0)

- Is it useful to pass an array (k,z) to the emulator rather than just an array of (k) for each z?

- Actually, do you prefer k in `Mpc^-1` or` h * Mpc^-1` ?

**Reply by gcanasherrera** [reply]
I CC @PedroCarrilho as our emulator expert.

**Reply by PedroCarrilho** [reply]
Hi @MatthieuSchaller, let me try to answer your questions:

- What we do now is to extrapolate in z and k with a power law above the emulated ranges. If you think this is a bad approach and have a better suggestion, that would be great. In general it is potentially easier for us if you return an error outside the range of the emulator, just in case we make a mistake somewhere which is easier to find it if the code gets an error instead of getting just an unexpected answer. 

- It would potentially be very useful to be able to pass a (k,z) array, in particular if that makes evaluating the emulator faster than running a for loop over the requested redshifts. It would also be potentially useful to return the emulated result in the k,z grid used internally by the emulator, without any interpolation, just so we can use the interpolations we have for everything else. But this is really a minor thing, so don't worry if this is not easily feasible.

- In terms of the units of k, we do use `Mpc^-1`, but this is easy enough to convert later on, so it isn't a problem really.

A separate question from me: what do you do when the emulator goes outside its trained parameter range? Is there an error or do you return an approximate result? For this, in other emulators we also use extrapolations (log-linear in this case), but I have a feeling this is not the most correct thing to do, so if you have any other suggestion (or just think we should not extrapolate at all) that would be useful to know.

Thanks!

**Comment by MatthieuSchaller** [comment]
For reference, this is what you get as a function of k and z for the default baryon physics parameters:

![fits_fid](https://github.com/user-attachments/assets/90cd89c2-a271-46c4-a20f-5811a3b98c0c)

In terms of k, the emulator was trained between the vertical dotted lines. In terms of z, we trained up to z=2.

 - At lower k than trained, I already simply return 1. 
 - At higher k than trained, I don't return anything (and could provide an error if it helps). Note that our suppression here is different from BACCO, BCemu, or HMcode (say) as they all (normally) reach a plateau at k>~5-10. i.e. they don't model the turn up due to halo contraction and cooling. Extrapolating to a constant beyond their range makes sense because they are already flat. For us, it would be strange given that at higher k than trained the function is expected to keep raising for a while.  The answer about what to do here is maybe related to how far in k you want to go. 
 - At higher z than trained I currently return an error. Note that I think I will extend the training range to z=3. At that point, the response is basically 1 everywhere. How far in z would you like data? And what behaviour is best for you beyond? Error? Return the same value as the last z bin? Or some extrapolation/interpolation towards response(k) = 1?
 
 For the baryon physics parameters, I currently do not restrict the range at all. If you run the emulator and ask it to also return an estimate of the error, then the variance in the extrapolated region just blows up telling you the emulator is really unsure. 
 I could also return some error beyond some range if that helps. Or make things saturate to a min/max in each dimension of the baryon physics.
 
Thanks for the info about Mpc^-1.  I did `h * Mpc^-1` as this is was everyone always plots but I did all the work without the `h`. Let me think whether it's better to kill it off entirely.

**Reply by PedroCarrilho** [reply]
Hi Matthieu,

Sorry for the delay in replying. 

- For low k, this is also what we have been assuming for all cases.

- For high k, I did have the worry at some point that this made a difference, but in the end it seems like it mostly does not matter what is done above say k=5 h/Mpc (the Bacco limit). As I mentioned I was extrapolating with a power law, and indeed for Bacco this is weird as it was not capturing the turn up as you say. Since your emulator goes up to k=30 h/Mpc I don't think we need to worry at all. We have been extrapolating other things up to k=500 Mpc^-1, but what we put there has not mattered. Still, I am curious, is there anything more realistic one could do there, i.e. is there some expectation that the response will have any asymptotic behaviour?

- Extending to z=3 would be perfect, as that is probably the furthest our n(z) will be non-negligible (probably around 2.5 already). If your expectation is that it is essentially =1 above that, then extrapolating to response=1 is great. I suppose what we have been doing so far is not the best option. 

- Regarding these and the parameter ranges, if you don't have it yet, it would be useful to have somewhere in the code a way for the user to access the ranges for all quantities. This allows us to know at what point we can access the prediction without extrapolation, even when no errors or warnings are returned. 

Regarding the parameter ranges, it sounds good to output something with the associated uncertainty. Potentially a warning might be useful, but I suppose you will tell people what the ranges are, so the user will know they are going beyond it at their own risk. In the figure above you show 3 parameters and then predictions at different z, so I am guessing the redshift evolution is predicted as well from those 3 parameters, right? In that case, most of these extrapolation issues are not so problematic, since the user can just run an analysis with a prior that matches the ranges of the emulator.

On the Mpc^-1, this is easy to convert, so if I was doing it, I might just leave it as it, since as you say it is what most people do, and I expect that your emulator may be used also beyond Euclid.

Beyond all of this, would it be possible for me to have access to the current version of the emulator to play with it a bit?

Thanks

**Reply by MatthieuSchaller** [reply]
I have added you to the repository. And will add a simple base example later today. Will likely have to break my vow of never using notebooks...

- for high k: OK. I can leave things as they are now. To answer your question, no, the behaviour at very high k is not known and unlikely to asymptote. At some point you will enter the real of the mass distribution inside galaxies themselves compare to the DMO behaviour that follows NFW. The question of what galaxy forms in which halo and with what density profile is far from settled. 

- I'll add z=3 then. I had assumed the Euclid n(z) and a lensing kernel peaking at ~1/2 of the distance when limiting things to z=2 but extending is easy.

- Good idea to add a function returning this. I had it in the docs but we may as well provide this as a service in the code.

Yes, we have the z evolution for any combination of the baryon parameter triplet.

**Reply by MatthieuSchaller** [reply]
Hi @PedroCarrilho,
I was wondering whether there was any progress and if there is anything I can help with here?

**Comment by PedroCarrilho** [comment]
Hi @MatthieuSchaller ,

Sorry for the delay in replying, it has been a tricky month. I have made a first implementation of the emulator, which seems to be working nicely. I did have an additional question though. In all other emulators the suppression depends on only one cosmological parameter - the baryon fraction. But FLAMINGO has no direct dependence on that. Is that because one of the emulated parameters is fully degenerate with that? If so, that would be useful to know. In other words, if one were to assume full knowledge of baryonic feedback physics, how would the background baryon fraction affect the suppression?

Thanks!

**Reply by MatthieuSchaller** [reply]
Hi @PedroCarrilho 

Great you got it to work!
Do you mean that the suppression in other model depends on Omega_b / Omega_m ? Or do you mean that it depends on observed baryon fractions in some objects?

**Reply by MatthieuSchaller** [reply]
In general, we find little dependence of the response on cosmological parameters. That's why the current version does not depend on any background cosmology info. Having said that, we have not explored wild changes in Omega_b / Omega_m as this ratio is already quite well constrained observationally.

**Reply by MatthieuSchaller** [reply]
Do you need a dependence on Omega_b/Omega_m or could the package happily assume there isn't one?

**Reply by PedroCarrilho** [reply]
Hi @MatthieuSchaller,

I meant that the suppression directly depends on f_b = Omega_b / Omega_m. I am aware that cosmology has a small role, but I would generally expect f_b to play a substantial role, since I don't expect any suppression if Omega_b ~ 0 and then it should more or less increase with Omega_b (for fixed Omega_m). 
For example BCemu depends on 7 feedback parameters plus f_b, and the effect is significant:
![image](https://github.com/user-attachments/assets/ce8c37d8-e004-4594-9fd9-0bee91edd29d)

As you say, f_b is fairly well constrained in LCDM and with non-LSS data. But for Euclid, we probably will have Euclid-only runs, in which f_b is not well constrained, in principle. So it would indeed be useful to know how the FLAMINGO emulator prediction changes with f_b, unless this would be completely degenerate with its other parameters. I can see this being true, and then we would not need any dependence on f_b, but it would probably be good to understand this better.

**Reply by MatthieuSchaller** [reply]
I see. So, that's not something we can currently provide. We don't have enough runs varying Omega_b / Omega_m to build an emulator.  

Though I agree that if you intend to vary the ratio as widely as on the plot above, there will be an impact.

**Reply by MatthieuSchaller** [reply]
Ok, thought a bit more about it and I can confirm that we can't provide a dependence on f_b at the moment or in the near future. 

Is it possible to issue warnings when things go beyond a certain range? Or to exclude this particular model when doing very large parameter searches?

**Reply by PedroCarrilho** [reply]
Hi Mathieu, yes, we can have warnings and we do have other models that we can use when we sample over a broad range of f_b. I suppose what would be good to confirm is what is the value of f_b for which Flamingo was trained. In one of the Flamingo papers, I found the following parameters Omega_m = 0.306, Omega_b = 0.0486, so that f_b = 0.158824. Can you confirm that?

**Reply by MatthieuSchaller** [reply]
That's correct. 

Note that in the paper we also show that the response is unchanged for other cosmologies we have like Planck (fb=0.15633) and the "lensing cosmology" of Amon& Efsathiou 2022 (fb=0.15508).

**Comment by MatthieuSchaller** [comment]
Apologies for the delays, but I have eventually added a `pyproject.toml` to the emulator's repo and also pushed it to `pypi`. 

I think this should now be enough to let you smoothly have the project as a dependency of cloe.
